File: frontend/src/lib/api/measurements.ts
import { api } from "./http";
import type {
  BodyMeasurement,
  BodyMeasurementInput,
  Measurement,
  MeasurementInput,
} from "@/types/measurement";

export function listMeasurements(clientId: string): Promise<Measurement[]> {
  return api.get<Measurement[]>(`/clients/${clientId}/measurements`);
}

export function createMeasurement(
  clientId: string,
  input: MeasurementInput,
): Promise<Measurement> {
  return api.post<Measurement>(`/clients/${clientId}/measurements`, input);
}

export function updateMeasurement(
  measurementId: string,
  input: Partial<MeasurementInput>,
): Promise<Measurement> {
  return api.patch<Measurement>(`/measurements/${measurementId}`, input);
}

export function deleteMeasurement(measurementId: string): Promise<void> {
  return api.delete<void>(`/measurements/${measurementId}`);
}

export function listBodyMeasurements(
  clientId: string,
): Promise<BodyMeasurement[]> {
  return api.get<BodyMeasurement[]>(`/clients/${clientId}/body-measurements`);
}

export function createBodyMeasurement(
  clientId: string,
  input: BodyMeasurementInput,
): Promise<BodyMeasurement> {
  return api.post<BodyMeasurement>(
    `/clients/${clientId}/body-measurements`,
    input,
  );
}

export function updateBodyMeasurement(
  measurementId: string,
  input: Partial<BodyMeasurementInput>,
): Promise<BodyMeasurement> {
  return api.patch<BodyMeasurement>(
    `/body-measurements/${measurementId}`,
    input,
  );
}

export function deleteBodyMeasurement(measurementId: string): Promise<void> {
  return api.delete<void>(`/body-measurements/${measurementId}`);
}

File: backend/app/api/measurements.py
import uuid

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from app.core.db import get_db
from app.core.security import get_current_trainer
from app.schemas.measurement import (
    BodyMeasurementCreate,
    BodyMeasurementOut,
    BodyMeasurementUpdate,
    MeasurementCreate,
    MeasurementOut,
    MeasurementUpdate,
)
from app.services import measurement_service

router = APIRouter(tags=["measurements"], dependencies=[Depends(get_current_trainer)])


@router.get(
    "/api/clients/{client_id}/measurements", response_model=list[MeasurementOut]
)
def list_measurements(
    client_id: uuid.UUID, db: Session = Depends(get_db)
) -> list[MeasurementOut]:
    return measurement_service.list_measurements(db, client_id)


@router.post(
    "/api/clients/{client_id}/measurements",
    response_model=MeasurementOut,
    status_code=status.HTTP_201_CREATED,
)
def create_measurement(
    client_id: uuid.UUID,
    payload: MeasurementCreate,
    db: Session = Depends(get_db),
) -> MeasurementOut:
    return measurement_service.create_measurement(db, client_id, payload)


@router.patch("/api/measurements/{measurement_id}", response_model=MeasurementOut)
def update_measurement(
    measurement_id: uuid.UUID,
    payload: MeasurementUpdate,
    db: Session = Depends(get_db),
) -> MeasurementOut:
    return measurement_service.update_measurement(db, measurement_id, payload)


@router.delete(
    "/api/measurements/{measurement_id}", status_code=status.HTTP_204_NO_CONTENT
)
def delete_measurement(
    measurement_id: uuid.UUID, db: Session = Depends(get_db)
) -> Response:
    measurement_service.delete_measurement(db, measurement_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get(
    "/api/clients/{client_id}/body-measurements",
    response_model=list[BodyMeasurementOut],
)
def list_body_measurements(
    client_id: uuid.UUID, db: Session = Depends(get_db)
) -> list[BodyMeasurementOut]:
    return measurement_service.list_body_measurements(db, client_id)


@router.post(
    "/api/clients/{client_id}/body-measurements",
    response_model=BodyMeasurementOut,
    status_code=status.HTTP_201_CREATED,
)
def create_body_measurement(
    client_id: uuid.UUID,
    payload: BodyMeasurementCreate,
    db: Session = Depends(get_db),
) -> BodyMeasurementOut:
    return measurement_service.create_body_measurement(db, client_id, payload)


@router.patch(
    "/api/body-measurements/{measurement_id}", response_model=BodyMeasurementOut
)
def update_body_measurement(
    measurement_id: uuid.UUID,
    payload: BodyMeasurementUpdate,
    db: Session = Depends(get_db),
) -> BodyMeasurementOut:
    return measurement_service.update_body_measurement(db, measurement_id, payload)


@router.delete(
    "/api/body-measurements/{measurement_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def delete_body_measurement(
    measurement_id: uuid.UUID, db: Session = Depends(get_db)
) -> Response:
    measurement_service.delete_body_measurement(db, measurement_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
